### PR TITLE
NIAD-824: create GitHub task to check branch and PR naming convention

### PR DESCRIPTION
**Pull request by saileshp-kainos**
Created a `pr-lint.yml` file to create GitHub Actions which contain two jobs that check
- the branch name meets the agreed convention
- the pull request name meets the agreed convention

### CODE REVIEW

File: .github/workflows/pr-lint.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+name: PR Quality Check
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, edited]
+jobs:
+  check-pr-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check ticket name conforms to requirements
+        run: 'echo ${{ github.event.pull_request.title }} | grep -i -E -q "^((apm|niad)-[0-9]+): ([a-zA-Z]+)|(dependabot\/)"'
```
Comment (bartek-sarul) [inline .github/workflows/pr-lint.yml:10]: `apm` won't be used here

File: .github/workflows/pr-lint.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+  check-branch-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check branch name confirms to requirements
```
Comment (bartek-sarul) [inline .github/workflows/pr-lint.yml:15]: conforms

File: .github/workflows/pr-lint.yml (hunk 1 above)
Comment (andrewwa-kainos) [inline .github/workflows/pr-lint.yml:10]: We can probably be confident we'll only be working in the NIAD project, so APM is not needed. Check with @bartek-sarul ?

Also the `-i` flag is "ignore case".
```
> man grep
/-i
```
```
    -i, --ignore-case
             Perform case insensitive matching.  By default, grep is case sensitive.
```
so the character set doesn't need to include both cases.
(This latter point applies to line 16 too.)

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-lint.yml (hunk 1 above)
Comment (saileshp-kainos) [reply .github/workflows/pr-lint.yml]: Change made

File: .github/workflows/pr-lint.yml (hunk 2 above)
Comment (saileshp-kainos) [reply .github/workflows/pr-lint.yml]: Spelling has been corrected

**Review by andrewwa-kainos — DISMISSED** [review]
I don't think you need all those capture groups in the regexes, but they're not hurting anything.

LGTM

How will you test it? Some github configs only take effect when they're in the main branch. If so, we might just have to merge it and see

**Comment by saileshp-kainos** [comment]
> I don't think you need all those capture groups in the regexes, but they're not hurting anything.
> 
> LGTM
> 
> How will you test it? Some github configs only take effect when they're in the main branch. If so, we might just have to merge it and see

@andrewwa-kainos so far, I've managed to test the pull request title using this pull request, but I've not been able to test the branch name check as that would require the GitHub Action to be on the branch in question

File: .github/workflows/pr-lint.yml (hunk 1 above)
Comment (antonlebedjko) [inline .github/workflows/pr-lint.yml]: If we are using the -i flag, don't think that A-Z is necessary in "[a-zA-Z]" part

File: .github/workflows/pr-lint.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        run: 'echo ${{ github.event.pull_request.head.ref }} | grep -i -E -q "(((feature|bug))/(niad)-[0-9]+-[a-zA-Z]+)"'
```
Comment (antonlebedjko) [inline .github/workflows/pr-lint.yml]: Same thing for A-Z here as on line 10

File: .github/workflows/pr-lint.yml [hunk 4]
```diff
@@ -7,10 +7,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Check ticket name conforms to requirements
-        run: 'echo ${{ github.event.pull_request.title }} | grep -i -E -q "^((apm|niad)-[0-9]+): ([a-zA-Z]+)|(dependabot\/)"'
+        run: 'echo ${{ github.event.pull_request.title }} | grep -i -E -q "^((niad)-[0-9]+): ([a-z]+)|(dependabot\/)"'
```
Comment (antonlebedjko) [inline .github/workflows/pr-lint.yml:10]: Are these "(" or ")" actually needed now? Also, doesn't your regex just accept one word? Cannot we have spaces between the words?

File: .github/workflows/pr-lint.yml [hunk 5]
```diff
@@ -7,10 +7,10 @@ jobs:
...
 
   check-branch-name:
     runs-on: ubuntu-latest
     steps:
-      - name: Check branch name confirms to requirements
-        run: 'echo ${{ github.event.pull_request.head.ref }} | grep -i -E -q "(((feature|bug))/(niad)-[0-9]+-[a-zA-Z]+)"'
+      - name: Check branch name conforms to requirements
+        run: 'echo ${{ github.event.pull_request.head.ref }} | grep -i -E -q "(((feature|bug))/(niad)-[0-9]+-[a-z]+)"'
```
Comment (antonlebedjko) [inline .github/workflows/pr-lint.yml:16]: We definitely have some extra round brackets which are not necessary

File: .github/workflows/pr-lint.yml (hunk 4 above)
Comment (andrewwa-kainos) [reply .github/workflows/pr-lint.yml:10]: It will accept more than one word after the JIRA key. See if you can work out why!

Comment (antonlebedjko) [reply .github/workflows/pr-lint.yml:10]: Oh, because we don't end with $. True

File: .github/workflows/pr-lint.yml (hunk 5 above)
Comment (saileshp-kainos) [reply .github/workflows/pr-lint.yml]: @antonlebedjko i've made these changes and removed the unnecessary brackets

File: .github/workflows/pr-lint.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+name: PR Quality Check
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, edited]
+jobs:
+  check-pr-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check ticket name conforms to requirements
+        run: 'echo ${{ github.event.pull_request.title }} | grep -i -E -q "^(niad)-[0-9]+: [a-z]+|(dependabot\/)"'
```
Comment (andrewwa-kainos) [inline .github/workflows/pr-lint.yml:10]: Now that it's simplified, we can also note that `dependabot/` also satifies `[a-z]+`, and is thus redundant. (Also, no need to capture `niad`.)
```suggestion
        run: 'echo ${{ github.event.pull_request.title }} | grep -i -E -q "^niad-[0-9]+: [a-z]+"'
```

File: .github/workflows/pr-lint.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+  check-branch-name:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check branch name conforms to requirements
+        run: 'echo ${{ github.event.pull_request.head.ref }} | grep -i -E -q "^(feature|bug)\/niad-[0-9]+-[a-z]+"'
```
Comment (andrewwa-kainos) [inline .github/workflows/pr-lint.yml:16]: Think you stripped out a bit too much 😁
We want to allow any positive number of _dash letters_ sequences, so we will need a group to represent that repeatable structure.
```suggestion
        run: 'echo ${{ github.event.pull_request.head.ref }} | grep -i -E -q "^(feature|bug)\/niad-[0-9]+(-[a-z]+)+"'
```

**Review by andrewwa-kainos — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-lint.yml (hunk 6 above)
Comment (saileshp-kainos) [reply .github/workflows/pr-lint.yml]: @andrewwa-kainos removed dependabot and capture group from regex

File: .github/workflows/pr-lint.yml (hunk 7 above)
Comment (saileshp-kainos) [reply .github/workflows/pr-lint.yml]: Made new capture group as suggested

**Review by andrewwa-kainos — APPROVED** [review]
LGTM

**Review by bartek-sarul — APPROVED** [review]

**Pull request merged**